packages/multichain-network-controller/src/MultichainNetworkController.ts

```diff
@@ -0,0 +1,204 @@
+import { BaseController } from '@metamask/base-controller';
+import { isEvmAccountType } from '@metamask/keyring-api';
+import type { InternalAccount } from '@metamask/keyring-internal-api';
+import type { NetworkClientId } from '@metamask/network-controller';
+import { isCaipChainId } from '@metamask/utils';
+
+import {
+  MULTICHAIN_NETWORK_CONTROLLER_METADATA,
+  getDefaultMultichainNetworkControllerState,
+} from './constants';
+import {
+  MULTICHAIN_NETWORK_CONTROLLER_NAME,
+  type MultichainNetworkControllerState,
+  type MultichainNetworkControllerMessenger,
+  type SupportedCaipChainId,
+} from './types';
+import {
+  checkIfSupportedCaipChainId,
+  getChainIdForNonEvmAddress,
+} from './utils';
+
+/**
+ * The MultichainNetworkController is responsible for fetching and caching account
+ * balances.
+ */
+export class MultichainNetworkController extends BaseController<
+  typeof MULTICHAIN_NETWORK_CONTROLLER_NAME,
+  MultichainNetworkControllerState,
+  MultichainNetworkControllerMessenger
+> {
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: MultichainNetworkControllerMessenger;
+    state?: Omit<
+      Partial<MultichainNetworkControllerState>,
+      'multichainNetworkConfigurationsByChainId'
+    >;
+  }) {
+    super({
+      messenger,
+      name: MULTICHAIN_NETWORK_CONTROLLER_NAME,
+      metadata: MULTICHAIN_NETWORK_CONTROLLER_METADATA,
+      state: {
+        ...getDefaultMultichainNetworkControllerState(),
+        ...state,
+      },
+    });
+
+    this.#subscribeToMessageEvents();
+    this.#registerMessageHandlers();
+  }
+
+  /**
+   * Sets the active EVM network.
+   *
+   * @param id - The client ID of the EVM network to set active.
+   */
+  async #setActiveEvmNetwork(id: NetworkClientId): Promise<void> {
+    // Notify listeners that setActiveNetwork was called
+    this.messagingSystem.publish(
+      'MultichainNetworkController:networkDidChange',
+      id,
+    );
```

Comment:
IMO this should have been fired after the actually update the network configuration.

If some logic is checking the actual state from this controller (and its counterpart the `NetworkController`) and react to this event, the state won't be "consistent", no?

Reply:
Updated